NOISSUE: Add kubeconfig management docs (#1520)
openshift/microshift: docs/howto_kubeconfig.md

# Kubeconfig management

> Every time MicroShift starts it generates a set of kubeconfig files for local and remote management of the apiserver. These are tied to the certificates for the apiserver and are also impacted by MicroShift's configuration.

Reviewer: It doesn't make new files every time, right? It updates them under some circumstances, but if they are present and valid it leaves them alone?

Author: File is created if it does not exist, but every time MicroShift starts it checks if it is up to date.
```bash
$ stat kubeconfig 
  File: kubeconfig
  Size: 5706      	Blocks: 16         IO Block: 4096   regular file
Device: fd00h/64768d	Inode: 33604994    Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:container_var_lib_t:s0
Access: 2023-03-21 08:18:42.601264403 +0000
Modify: 2023-03-21 08:18:35.564777086 +0000
Change: 2023-03-21 08:18:35.564777086 +0000
 Birth: 2023-03-17 08:44:10.201752721 +0000
$ systemctl restart microshift
$ stat kubeconfig 
  File: kubeconfig
  Size: 5706      	Blocks: 16         IO Block: 4096   regular file
Device: fd00h/64768d	Inode: 33604994    Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:container_var_lib_t:s0
Access: 2023-03-21 08:25:28.343826507 +0000
Modify: 2023-03-21 08:25:21.306804581 +0000
Change: 2023-03-21 08:25:21.306804581 +0000
 Birth: 2023-03-17 08:44:10.201752721 +0000
```

> Every time MicroShift starts it generates a set of kubeconfig files for local and remote management of the apiserver. These are tied to the certificates for the apiserver and are also impacted by MicroShift's configuration.

Reviewer 2: ```suggestion
Every time MicroShift starts, it checks if the set of kubeconfig files for local and remote management of the apiserver exist and are up-to-date. These are tied to the certificates for the apiserver and are also impacted by MicroShift's configuration.
```

## Configuration
MicroShift allows configuring additional names and IP addresses on top of the defaults, which are `<node.hostnameOverride>`, `<node.nodeIP>` and `api.<dns.baseDomain>`. A configuration example:
```yaml
dns:
  baseDomain: example.com
node:
  hostnameOverride: "microshift-rhel9"
  nodeIP: 10.0.0.1
apiServer:
  subjectAltNames:
  - alt-name-1
  - 1.2.3.4
```
All these parameters are included as common names (CN) and subject alternative names (SAN) in the external serving certificates for apiserver.

## Kubeconfig files

> Upon starting, MicroShift generates a set of kubeconfig files under `/var/lib/microshift/resources/kubeadmin` for different network access. From the example above we would have:

Reviewer 2: ```suggestion
Upon starting, MicroShift generates a set of kubeconfig files under `/var/lib/microshift/resources/kubeadmin` for different network access types. From the example above we would have:
```

```bash
/var/lib/microshift/resources/kubeadmin/
├── kubeconfig
├── alt-name-1
│   └── kubeconfig
├── 1.2.3.4
│   └── kubeconfig
└── microshift-rhel9
    └── kubeconfig
```
Sections below will assume this configuration is loaded into the system.

### Local access

> When connecting to the apiserver locally, this certificate is served for validation:

Author: Done

```bash
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = kube-apiserver-localhost-signer
        Validity
            Not Before: Mar 17 08:44:07 2023 GMT
            Not After : Mar 16 08:44:08 2024 GMT
        Subject: CN = localhost
...
            X509v3 Subject Alternative Name: 
                DNS:localhost
```
This means we need to use `localhost` to connect to it, and have our kubeconfig use the same CA that signed the certificate above.

The local access kubeconfig is always generated at `/var/lib/microshift/resources/kubeadmin/kubeconfig`. This one is not impacted/driven by configuration parameters. It allows access to apiserver using `localhost` and uses the `localhost` internal CA:
```yaml
clusters:
- cluster:
    certificate-authority-data: <base64 CA>
    server: https://localhost:6443
```

> This file can not be used outside of the MicroShift's host, as the CA it uses validates `localhost` network access served certificates only.

Reviewer 2: ```suggestion
This file cannot be used outside of the MicroShift host as its CA only validates certificates with `localhost` network access type.
```

### External access

> When connecting to the apiserver from an external source this certificate is served for validation:

Reviewer: ```suggestion
When connecting to the apiserver from an external source a certificate like this one, with all of the other names in the SAN field, is served for validation:
```

Author: Done

> When connecting to the apiserver from an external source a certificate like this one, with all of the other names in the SAN field, is served for validation:

Reviewer 2: ```suggestion
When connecting to the API server from an external source, a certificate with all of the alternative names in the SAN field is served for validation, as in the following example.
```

Author: Done. Changed all references to apiserver in the doc too.

```bash
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = kube-apiserver-external-signer
        Validity
            Not Before: Mar 17 08:44:07 2023 GMT
            Not After : Mar 16 08:44:08 2024 GMT
        Subject: CN = 10.0.0.1
...
            X509v3 Subject Alternative Name: 
                DNS:api.example.com, DNS:microshift-rhel9, DNS:alt-name-1, DNS:10.0.0.1, DNS:1.2.3.4, IP Address:10.0.0.1, IP Address:1.2.3.4
```
As with the local access use case, we need the kubeconfig to include one of those names as `server` and also use the same CA for validation.

MicroShift will generate a default kubeconfig for external access using the hostname, and an additional one per entry in `apiServer.subjectAltNames`.

> `/var/lib/microshift/resources/kubeadmin/microshift-rhel9/kubeconfig` uses the machine's hostname (`node.hostnameOverride` or the default hostname value) to reach apiserver and its CA is able to validate external access served certificates.

Reviewer: ```suggestion
`/var/lib/microshift/resources/kubeadmin/microshift-rhel9/kubeconfig` uses the machine's hostname (or `node.hostnameOverride`, if that configuration option is set) to reach the apiserver and its CA is able to validate external access served certificates.
```

Author: Done

> `/var/lib/microshift/resources/kubeadmin/microshift-rhel9/kubeconfig` uses the machine's hostname (or `node.hostnameOverride` if that option is set) to reach apiserver and its CA is able to validate external access served certificates.

Reviewer 2: ```suggestion
`/var/lib/microshift/resources/kubeadmin/<hostname>/kubeconfig` uses the machine's host name (or `node.hostnameOverride` if that option is set) to reach the API server. Its CA is able to validate certificates when accessed externally.
```

Author: Done.
Keeping hostname's spelling: https://en.wikipedia.org/wiki/Hostname

```yaml
clusters:
- cluster:
    certificate-authority-data: <base64 CA>
    server: https://microshift-rhel9:6443
```

> As we have seen in the served certificate we can swap the hostname in `server` stanza to use the node IP instead.

Reviewer 2: ```suggestion
> As we have seen in the served certificate, we can swap the hostname in `server` stanza to use the node IP instead.
```

Author: Done

> `/var/lib/microshift/resources/kubeadmin/alt-name-1/kubeconfig` and `/var/lib/microshift/resources/kubeadmin/1.2.3.4/kubeconfig` come from the `apiServer.subjectAltNames` configuration values. These are useful to refer to the MicroShift's host using different names than their hostname and/or IPs. For example, using a different DNS name or an external IP.

Reviewer 2: ```suggestion
`/var/lib/microshift/resources/kubeadmin/<alt-name-1>/kubeconfig` and `/var/lib/microshift/resources/kubeadmin/<1.2.3.4>/kubeconfig` come from the `apiServer.subjectAltNames` configuration values. These are useful to refer to the MicroShift's host using alternative host names and/or IPs (i.e. using a different DNS name or an external IP).
```

Author: Done
